File: solutions/free-pascal/isbn-verifier/1/IsbnVerifier.pas
unit IsbnVerifier;

{$mode ObjFPC}{$H+}

interface

function IsValid(const isbn : string) : boolean;

implementation

uses SysUtils;

function IsValid(const isbn : string) : boolean;
var 
  sum : integer;
  factor : integer;
  ch : char;
  val : integer;
  valid : boolean;
begin
  sum := 0;
  factor := 10;
  for ch in isbn do
  begin
    if ch='-' then
      continue;
      
    valid := ((ord(ch)>=ord('0')) and (ord(ch)<=ord('9'))) 
      or ((ch='X') and (factor=1));
    if not valid then
      exit(false);
      
    if ch='X' then 
      val := 10
    else
      val := ord(ch) - ord('0');
    sum := sum + val*factor;
    dec(factor);
  end;
  result := (factor=0) and (sum mod 11 = 0);
end;

end.

File: solutions/cpp/isbn-verifier/3/isbn_verifier.cpp
#include "isbn_verifier.h"

#include <vector>

using namespace std;

namespace isbn_verifier {

    bool is_valid(string isbn) {
        unsigned int factor=10;
        int sum = 0;
        for(auto ch : isbn) {
            if (ch=='-') 
                continue;            
                        
            if (isdigit(ch) || (ch=='X' && factor==1)) {
                sum += factor*(ch == 'X' ? 10 : ch - '0');
            }    
            else  
                return false;
            factor--;
        }
        return factor==0 && 0==sum%11;
    }


}  // namespace isbn_verifier